### ISSUE TITLE
First API ideas

### ISSUE THREAD

**Issue by BraisGabin** [PoGoTools/PoGoTools#4]
This is just the first proposal and it's more less what I have on PokeScreenshot.

Two classes with two different responsabilities (please, help me with the naming). Each class only works with a single screenshot. You discard these objects as soon as the processing is done. Then you generates new ones for the next screenshot.
# `Tess`

It's the only objects that knows about Tesseract. Responsabilties:
- It has the Bitmap (it doesn't create it, but it is the one who keeps the reference to it).
- It does the preprocessing. Color Bitmap to Black&White.
- It knows the positions of the different elements: coordinates of CP, HP, Candy...
- It doesn't have any state.
- It just returns `String`s. It doesn't know if that String have any sense.
# `ScreenshotReader`

It's the one that talks with Tess. And it knows the meaning of each value. it knows, for example, the correct values of startdust. Responsabilies:
- It return real data. I mean, the CP as int, the candy as a known candy name, etc.
  - This means that he knows how to clean the _dirty_ Strings from `Tess`. This is the heuristics' place.
- It fails when the data isn't correct (I don't know the way, see #3).
- It have state. It remembers the values. If you asks twice for the CP it doesn't calculate it again.

Possible API:

``` java
public interface Tess {
  String cp();
  String name();
  String hp();
  String weight();
  String type();
  String height();
  String totalStardust();
  String candy();
  String startdust();
  String candyCost();
}
```

``` java
public interface ScreenshotReader {
  int cp();
  String name();
  int hp();
  float weight();
  // We provide a matrix. Each row has all the valid translations of each type
  String[] type(String[][]);
  float height();
  int totalStardust();
  // We provide a matrix. Each row has all the valid translations of each Pokémon "branch"
  String[] candy(String[][] validCandyNames);
  // I'm not sure with this one... I think that it's more usefull as String than int
  int startdust(int[] validStardustValues);
  int candyCost();
}
```

**Comment by Blaisorblade** [comment]
I can be fine with separating raw OCR and the ScreenshotReader. But for GoIV, as discussed in #3, we need some result object that can represent partially failed scans. Maybe a third `ScreenshotData`?

BTW, have you looked at OCRHelper?

Also, regarding these fields:

> // We provide a matrix. Each row has all the valid translations of each Pokémon "branch"
>   String[] candy(String[][] validCandyNames);
> 
> // We provide a matrix. Each row has all the valid translations of each type
>   String[] type(String[][]);

It took me a longer while to get that, but if it is what I think, `String[][]` seems an overly restrictive interface (compare with the more general `List<List<String>>`). 

Re stardust: I'd prefer it as an integer frankly, since all valid values can be represented as both but string allows more invalid values.

**Comment by Blaisorblade** [comment]
Another note: some code somewhere (in the library or not) will end up needing a `PokemonData` _without_ any Optional — like what GoIV produces after the user inputs missing data, or what pokescreenshot gets in the first place.

That has a similar interface to `ScreenshotReader`, but 
- `type` etc. are just values, not functions
- this is really a `PokemonData`, not `ScreenshotReader`.

Additionally, and relatedly, catching `candy`, `stardust` or `type` in `ScreenshotReader` is tricky — you can easily cache them if you force the caller to always pass the same arguments (or check that this is done), but then it's easier to just store the input constants in some fixed location—some code (not in OCR) will probably need game constants anyway.

I've now noticed the return types of `candy`, `type`, etc. I'm not sure why this API supports multiple translation at once: GoIV mostly uses only a single locale (the active one). Even more importantly, why return a row instead of an appropriate ID for the type or the Pokemon?

Also: how much do we want to get only some data without others?

**Comment by BraisGabin** [comment]
> But for GoIV, as discussed in #3, we need some result object that can represent partially failed scans. Maybe a third ScreenshotData?

I think that ScreenshotReader is ScreenshotData, isn't it? I mean, It caches all the data so you if you needs ALL the data you can get it from it. If you need some value twice, it's ok, you can take it easily and fast. And, what it you don't need some value? If you read `Mr. Mine` as candy, you don't need the candy cost. And if you want to use it just to be sure that it's a real Mr. Mime you can do it too.

> It took me a longer while to get that, but if it is what I think, String[][] seems an overly restrictive interface (compare with the more general List<List<String>>).

You are right. We can use List<List<String>>.

> Re stardust: I'd prefer it as an integer frankly, since all valid values can be represented as both but string allows more invalid values.

👍 

> Another note: some code somewhere (in the library or not) will end up needing a PokemonData without any Optional — like what GoIV produces after the user inputs missing data, or what pokescreenshot gets in the first place.

You are right. In my code I have this solution (as always, just an idea):

``` java
public interface CoreStats {
  int atk();
  int def();
  int stam();
}
```

My guesser needs this values. So in the API I demand a CoreStats object. I don't bother about the real type. The real implementation is a sqlite row. But any implementation is possible.

> Additionally, and relatedly, catching candy, stardust or type in ScreenshotReader is tricky — you can easily cache them if you force the caller to always pass the same arguments (or check that this is done), but then it's easier to just store the input constants in some fixed location—some code (not in OCR) will probably need game constants anyway.

You are completely right. We can't cache values if we have arguments... We probably can add all that dependecies in the constructor. Somthing like (the naming is awful): 

``` java
class ImpScreenshotReader implements ScreenshotReader {
  public ScreenshotReader create() {
    return new ImpScreenshotReader(Candy.VALUES, Type.VALUES, Stardust.VALUES);
  public 
    return ImpScreenshotReader(List<List<String>> candies, List<List<<String>> types, List<Integer> stardust) {
    // ...
  }
  }
}
```

We have our default values if you use the handy `create` function. You can use whatever you want if you use the constructor.

The types can be a bit tricky. But, as long as GoIV and Pokescreenshot doesn't use it, we can just expose the API and throw a friendly `UnsupportedOperationException("We'll be happy if you want to PR this feature :)")`

> I've now noticed the return types of candy, type, etc. I'm not sure why this API supports multiple translation at once: GoIV mostly uses only a single locale (the active one). Even more importantly, why return a row instead of an appropriate ID for the type or the Pokemon?

I must say that I don't know how to fix this problem in a good way. I understand your point, List<String> as a candy value sucks. But I don't like the idea to have the pokemon id in the library... Somthing like?

``` java
interface Candy {
  String candy() { // this is the value that 99% people need
    return candy("en");
  }

  String candy(String locale) {
    return candyNames.get(locale); // Just to show the idea, it's not a real implementation
  }

  List<String> allNames() {
    return Arrays.asList(candy("en"), candy("fr"), candy("de"));
  }
}
```

I like the idea of parse all the languages at once. An use case example: https://github.com/farkam135/GoIV/issues/463

**Comment by Blaisorblade** [comment]
> I think that ScreenshotReader is ScreenshotData, isn't it?

Kind of, except for the functions taking arguments, and the name—one name suggests you're parsing a screenshot, the other isn't. But once we agree on the role, the name is nitpicking.
But let's say we fix those to never take arguments as you're also suggesting.

However, if an interface has only results, its methods shouldn't throw exceptions. But methods in `ScreenshotReader` throw instead, and that affects the interface. So, still oversimplifying, I might end up with:

``` java
public interface ScreenshotReader {
  String name(); // Can *this* ever throw? Think not.
  int cp() throws ScanFailedException;
  int hp() throws ScanFailedException;
  float weight() throws ScanFailedException;
  String[] type() throws ScanFailedException;
  float height() throws ScanFailedException;
  int totalStardust() throws ScanFailedException;
  String[] candy() throws ScanFailedException;
  int startdust() throws ScanFailedException;
  int candyCost() throws ScanFailedException;
}

public interface ScreenshotData {
  int cp();
  String name();
  int hp();
  float weight();
  String[] type();
  float height();
  int totalStardust();
  String[] candy();
  int startdust();
  int candyCost();
}
```

That's on top of the variant with `Optional` (which, however, might not need both an interface and an implementation).

**Comment by BraisGabin** [comment]
I see your point and I agree, they are different. And I think that never take arguments is a good decision. 

Leaving the naming problem. Can the optional implementation of ScreenshorReader be the screenshotData? I don't want to return -1 as an error value.

**Comment by Blaisorblade** [comment]
> Can the optional implementation of ScreenshorReader be the screenshotData? I don't want to return -1 as an error value.

Ah, you mean replace the `ScreenshotData` I proposed with `ScreenshotDataOptional`? I intended `ScreenshotData` to be only used for complete screenshots which don't have missing values. Downside: to have accurate types, there are three variants of the same interface with subtle differences, and I don't know how to avoid that in Java (except by _not_ having accurate types).

**Comment by BraisGabin** [comment]
You are right, there isn't a way to do that without mayn interfaces... I think that we can afford that downside.

Ok, so we have, more less, the first idea. I'll start with an implementation and a PR to talk about real code :).

**Comment by BraisGabin** [comment]
8 days and no updates. Sorry for that. New job and a lot of work. I just want to say that I'm still here and I'll create that PR soon. I promise!
